File: src/python/gudhi/delaunay_complex.py
# This file is part of the Gudhi Library - https://gudhi.inria.fr/ - which is released under MIT.
# See file LICENSE or go to https://gudhi.inria.fr/licensing/ for full license details.
# Author(s):  Vincent Rouvreau
#
# Copyright (C) 2016 Inria
#
# Modification(s):
#   - YYYY/MM Author: Description of the modification

__author__ = "Vincent Rouvreau"
__maintainer__ = "Thibaud Kloczko"
__copyright__ = "Copyright (C) 2016 Inria"
__license__ = "MIT"

import warnings

from gudhi import _delaunay_complex_ext as t

from typing import Literal, Optional
from collections.abc import Iterable

from gudhi.simplex_tree import SimplexTree


# DelaunayComplex python interface
class DelaunayComplex(t.Delaunay_complex_interface):
    """DelaunayComplex is a simplicial complex constructed from the finite cells of a Delaunay Triangulation.

    When :paramref:`~gudhi.DelaunayComplex.create_simplex_tree.filtration` is:

    * `None` (default value) - The filtration value of each simplex is not computed (set to `NaN`)
    * `'alpha'`              - The filtration value of each simplex is computed as an :class:`~gudhi.AlphaComplex`
    * `'cech'`               - The filtration value of each simplex is computed as a :class:`~gudhi.DelaunayCechComplex`
    """

    def __init__(
        self,
        points: Iterable[Iterable[float]] = [],
        weights: Optional[Iterable[float]] = None,
        precision: Literal["fast", "safe", "exact"] = "safe",
    ):
        """Args:
              points (Iterable[Iterable[float]]): A list of points in d-Dimension.
              weights (Optional[Iterable[float]]): A list of weights. If set, the number of weights must correspond to
              the number of points.
              precision (str): Complex precision can be `'fast'`, `'safe'` or `'exact'`. Default is `'safe'`.

        :raises ValueError: In case of inconsistency between the number of points and weights.
        """
        assert precision in [
            "fast",
            "safe",
            "exact",
        ], "Delaunay complex precision can only be 'fast', 'safe' or 'exact'"
        fast = precision == "fast"
        exact = precision == "exact"

        # weights are set but is inconsistent with the number of points
        if weights is not None and len(weights) != len(points):
            raise ValueError("Inconsistency between the number of points and weights")
        else:
            if weights is None:
                weights = []

        super().__init__(points, weights, fast, exact)

    def create_simplex_tree(
        self,
        max_alpha_square: float = float("inf"),
        filtration: Optional[Literal["alpha", "cech"]] = None,
        output_squared_values: bool = True,
    ) -> SimplexTree:
        """
        Args:
            max_alpha_square (float): The maximum alpha square threshold the simplices shall not exceed. Default is set to
                                      infinity, and there is very little point using anything else since it does not save time.
            filtration (Literal): Set this value to `None` (default value) if filtration values are not needed to be computed
                                  (will be set to `NaN`). Set it to `alpha` to compute the filtration values with the Alpha complex, or
                                  to `cech` to compute the Delaunay Cech complex.
            output_squared_values (bool): Square filtration values when `True`. Default is `True`.
        Returns:
            SimplexTree: A simplex tree created from the Delaunay Triangulation. The vertex `k` corresponds to the k-th
                         input point. The vertices may not be numbered contiguously as some points may be discarded in the
                         triangulation (duplicate points, weighted hidden point, ...).
        """
        if not filtration in [None, "alpha", "cech"]:
            raise ValueError(
                f"'{filtration}' is not a valid filtration value. Must be None, 'alpha' or 'cech'"
            )

        filt = t.Filtration.NONE
        if filtration == "cech":
            filt = t.Filtration.CECH
        elif filtration == "alpha":
            filt = t.Filtration.ALPHA

        stree = SimplexTree()
        super().create_simplex_tree(stree, max_alpha_square, filt, output_squared_values)
        return stree


class AlphaComplex(DelaunayComplex):
    """AlphaComplex is a simplicial complex constructed from the finite cells of a Delaunay Triangulation.

    The filtration value of each simplex is computed as the squared radius of the smallest empty sphere passing through
    all of its vertices.

    All simplices that have a filtration value strictly greater than a given alpha squared value are not inserted into
    the complex.

    For more details about the algorithm, please refer to the
    `Alpha complex C++ documentation <https://gudhi.inria.fr/doc/latest/group__alpha__complex.html>`_

    .. note::

        When DelaunayComplex is constructed with an infinite value of alpha, the complex is a Delaunay complex.
    """

    def create_simplex_tree(
        self,
        max_alpha_square: float = float("inf"),
        default_filtration_value: bool = False,
        output_squared_values: bool = True,
    ) -> SimplexTree:
        """
        Args:
            max_alpha_square (float): The maximum alpha square threshold the simplices shall not exceed. Default is set to
                                      infinity, and there is very little point using anything else since it does not save time.
            default_filtration_value (bool): Default value is `False` (which means compute the filtration values). Set this
                                             value to `True` if filtration values are not needed to be computed (will be set to `NaN`), but please
                                             consider constructing a :class:`~gudhi.DelaunayComplex` instead.
            output_squared_values (bool): Square filtration values when `True`. Default is `True` to keep backward
                                          compatibility.
        Returns:
            SimplexTree: A simplex tree created from the Delaunay Triangulation. The vertex `k` corresponds to the k-th
                         input point. The vertices may not be numbered contiguously as some points may be discarded in the
                         triangulation (duplicate points, weighted hidden point, ...).
        """
        filtration = "alpha"
        if default_filtration_value:
            filtration = None
            warnings.warn(
                """Since Gudhi 3.10, creating an AlphaComplex with default_filtration_value=True is deprecated.
                          Please consider constructing a DelaunayComplex instead.
                          """,
                DeprecationWarning,
            )
        return super().create_simplex_tree(max_alpha_square, filtration, output_squared_values)


class DelaunayCechComplex(DelaunayComplex):
    """DelaunayCechComplex is a simplicial complex constructed from the finite cells of a Delaunay Triangulation.

    The filtration value of each simplex is equal to the squared radius of its minimal enclosing ball (MEB).

    All simplices that have a filtration value strictly greater than a given alpha squared value are not inserted into
    the complex.

    .. note::

        When DelaunayCechComplex is constructed with an infinite value of alpha, the complex is a Delaunay complex.
    """

    def __init__(self, points=[], precision="safe"):
        """
        Args:
            points (Iterable[Iterable[float]]): A list of points in d-Dimension.
            precision (str): Complex precision can be `'fast'`, `'safe'` or `'exact'`. Default is `'safe'`.
        """
        super().__init__(points=points, weights=None, precision=precision)

    def create_simplex_tree(
        self, max_alpha_square: float = float("inf"), output_squared_values: bool = True
    ) -> SimplexTree:
        """
        Args:
            max_alpha_square (float): The maximum alpha square threshold the simplices shall not exceed. Default is set to
                                      infinity, and there is very little point using anything else since it does not save time.
            output_squared_values (bool): Square filtration values when `True`. Default is `True`.
        Returns:
            SimplexTree: A simplex tree created from the Delaunay Čech Triangulation. The vertex `k` corresponds to the
                         k-th input point. The vertices may not be numbered contiguously as some points may be discarded in the
                         triangulation(duplicate points, ...).
        """
        return super().create_simplex_tree(max_alpha_square, "cech", output_squared_values)


#
# delaunay_complex.py ends here

File: src/python/gudhi/_delaunay_complex.cc
/*    This file is part of the Gudhi Library - https://gudhi.inria.fr/ - which is released under MIT.
 *    See file LICENSE or go to https://gudhi.inria.fr/licensing/ for full license details.
 *    Author(s):       Vincent Rouvreau
 *
 *    Copyright (C) 2020 Inria
 *
 *    Modification(s):
 *      - 2024/03 Vincent Rouvreau: Renamed Alpha_complex_factory as Delaunay_complex_factory for DelaunayCechComplex.
 *                                  Factorize create_complex
 *      - 2024/10 Vincent Rouvreau: Add square root filtration values interface
 *      - 2025/03 Thibaud Kloczko: Use nanobind instead of Cython for python bindings
 *      - 2025/04 Hannah Schreiber: Re-add possibility of tensors (numpy, torch etc.) as input
 *      - YYYY/MM Author: Description of the modification
 */

#include <vector>
#include <memory>   // for std::unique_ptr
#include <cstddef>  // for std::size_t

#include <nanobind/nanobind.h>
#include <nanobind/stl/string.h>
#include <nanobind/stl/vector.h>

#include <python_interfaces/Simplex_tree_interface.h>
#include <python_interfaces/points_utils.h>
#include <python_interfaces/delaunay_complex_interface.h>

namespace Gudhi {
namespace delaunay_complex {

std::unique_ptr<Abstract_delaunay_complex> make_2d_epick_complex_ptr(const Sequence2D&, const Sequence1D&, bool);
std::unique_ptr<Abstract_delaunay_complex> make_2d_epeck_complex_ptr(const Sequence2D&, const Sequence1D&, bool);
std::unique_ptr<Abstract_delaunay_complex> make_2d_epick_complex_ptr(const Sequence2D&, bool);
std::unique_ptr<Abstract_delaunay_complex> make_2d_epeck_complex_ptr(const Sequence2D&, bool);
std::unique_ptr<Abstract_delaunay_complex> make_3d_epick_complex_ptr(const Sequence2D&, const Sequence1D&, bool);
std::unique_ptr<Abstract_delaunay_complex> make_3d_epeck_complex_ptr(const Sequence2D&, const Sequence1D&, bool);
std::unique_ptr<Abstract_delaunay_complex> make_3d_epick_complex_ptr(const Sequence2D&, bool);
std::unique_ptr<Abstract_delaunay_complex> make_3d_epeck_complex_ptr(const Sequence2D&, bool);
std::unique_ptr<Abstract_delaunay_complex> make_dynamic_d_epick_complex_ptr(const Sequence2D&, const Sequence1D&, bool);
std::unique_ptr<Abstract_delaunay_complex> make_dynamic_d_epeck_complex_ptr(const Sequence2D&, const Sequence1D&, bool);
std::unique_ptr<Abstract_delaunay_complex> make_dynamic_d_epick_complex_ptr(const Sequence2D&, bool);
std::unique_ptr<Abstract_delaunay_complex> make_dynamic_d_epeck_complex_ptr(const Sequence2D&, bool);

class Delaunay_complex_interface
{
 public:
  Delaunay_complex_interface(const Sequence2D& points, const Sequence1D& weights, bool fast_version, bool exact_version)
  {
    // Specific cases for dimensions 2 and 3
    const std::size_t dimension = ((points.size() > 0) ? points[0].size() : 0);
    if (fast_version) {
      if (dimension == 2) {
        delaunay_ptr_ = make_2d_epick_complex_ptr(points, weights, exact_version);
      } else if (dimension == 3) {
        delaunay_ptr_ = make_3d_epick_complex_ptr(points, weights, exact_version);
      } else {
        delaunay_ptr_ = make_dynamic_d_epick_complex_ptr(points, weights, exact_version);
      }
    } else {
      if (dimension == 2) {
        delaunay_ptr_ = make_2d_epeck_complex_ptr(points, weights, exact_version);
      } else if (dimension == 3) {
        delaunay_ptr_ = make_3d_epeck_complex_ptr(points, weights, exact_version);
      } else {
        delaunay_ptr_ = make_dynamic_d_epeck_complex_ptr(points, weights, exact_version);
      }
    }
  }

  Delaunay_complex_interface(const Tensor2D& points, const Tensor1D& weights, bool fast_version, bool exact_version)
      : Delaunay_complex_interface(_get_sequence_from_tensor(points),
                                   _get_sequence_from_tensor(weights),
                                   fast_version,
                                   exact_version)
  {}

  Delaunay_complex_interface(const Sequence2D& points, bool fast_version, bool exact_version)
  {
    // Specific cases for dimensions 2 and 3
    const std::size_t dimension = ((points.size() > 0) ? points[0].size() : 0);
    if (fast_version) {
      if (dimension == 2) {
        delaunay_ptr_ = make_2d_epick_complex_ptr(points, exact_version);
      } else if (dimension == 3) {
        delaunay_ptr_ = make_3d_epick_complex_ptr(points, exact_version);
      } else {
        delaunay_ptr_ = make_dynamic_d_epick_complex_ptr(points, exact_version);
      }
    } else {
      if (dimension == 2) {
        delaunay_ptr_ = make_2d_epeck_complex_ptr(points, exact_version);
      } else if (dimension == 3) {
        delaunay_ptr_ = make_3d_epeck_complex_ptr(points, exact_version);
      } else {
        delaunay_ptr_ = make_dynamic_d_epeck_complex_ptr(points, exact_version);
      }
    }
  }

  Delaunay_complex_interface(const Tensor2D& points, bool fast_version, bool exact_version)
      : Delaunay_complex_interface(_get_sequence_from_tensor(points), fast_version, exact_version)
  {}

  std::vector<double> get_point(int vh) { return delaunay_ptr_->get_point(vh); }

  void create_simplex_tree(Simplex_tree_interface* simplex_tree,
                           double max_alpha_square,
                           Delaunay_filtration filtration,
                           bool output_squared_values)
  {
    // Nothing to be done in case of an empty point set
    if (delaunay_ptr_->num_vertices() > 0) {
      delaunay_ptr_->create_simplex_tree(simplex_tree, max_alpha_square, filtration, output_squared_values);
    }
  }

  static void set_float_relative_precision(double precision)
  {
    // cf. CGAL::Epeck_d kernel type in Delaunay_complex_interface
    if (precision <= 0 || precision >= 1) {
      throw std::invalid_argument("Precision must be strictly greater than 0 and lower than 0");
    }
    CGAL::Epeck_d<CGAL::Dynamic_dimension_tag>::FT::set_relative_precision_of_to_double(precision);
  }

  static double get_float_relative_precision()
  {
    // cf. CGAL::Epeck_d kernel type in Delaunay_complex_interface
    return CGAL::Epeck_d<CGAL::Dynamic_dimension_tag>::FT::get_relative_precision_of_to_double();
  }

 private:
  std::unique_ptr<Abstract_delaunay_complex> delaunay_ptr_;
};

}  // namespace delaunay_complex
}  // namespace Gudhi

// /////////////////////////////////////////////////////////////////////////////
// Delaunay_complex_interface wrapping
// /////////////////////////////////////////////////////////////////////////////

namespace nb = nanobind;
namespace gdc = Gudhi::delaunay_complex;
using gdci = gdc::Delaunay_complex_interface;

NB_MODULE(_delaunay_complex_ext, m)
{
  m.attr("__license__") = "GPL v3";

  nb::enum_<gdc::Delaunay_filtration>(m, "Filtration")
      .value("NONE", gdc::Delaunay_filtration::NONE, "Default Delaunay Complex")
      .value("CECH", gdc::Delaunay_filtration::CECH, "Delaunay Cech Complex")
      .value("ALPHA", gdc::Delaunay_filtration::ALPHA, "Alpha Complex");

  nb::class_<gdci>(m, "Delaunay_complex_interface")
      .def(nb::init<const Sequence2D&, const Sequence1D&, bool, bool>(), nb::call_guard<nb::gil_scoped_release>())
      .def(nb::init<const Tensor2D&, const Tensor1D&, bool, bool>(), nb::call_guard<nb::gil_scoped_release>())
      .def(nb::init<const Sequence2D&, bool, bool>(), nb::call_guard<nb::gil_scoped_release>())
      .def(nb::init<const Tensor2D&, bool, bool>(), nb::call_guard<nb::gil_scoped_release>())
      .def("create_simplex_tree", &gdci::create_simplex_tree, nb::call_guard<nb::gil_scoped_release>())
      .def("get_point", &gdci::get_point, R"doc(
This function returns the point corresponding to a given vertex from the :class:`~gudhi.SimplexTree` (the
same as the k-th input point, where `k=vertex`)

Args:
    vertex: The vertex.
Returns:
    the point.

:raises IndexError: In case the point has no associated vertex in the diagram (because of weights or because it
    is a duplicate).
        )doc")
      .def_static("get_float_relative_precision", &gdci::get_float_relative_precision, R"doc(
Get the float relative precision of filtration values computation when constructing with :code:`precision = 'safe'`
(the default).

Returns:
    The float relative precision.
        )doc")
      .def_static("set_float_relative_precision", &gdci::set_float_relative_precision, R"doc(
Set the float relative precision of filtration values computation when constructing with :code:`precision = 'safe'`
(the default).

Args:
    precision: When constructing :func:`~gudhi.delaunay_cech_complex`, :func:`~gudhi.alpha_complex`, or
        :func:`~gudhi.weighted_alpha_complex` with :code:`precision = 'safe'` (the default), one can
        set the float relative precision of filtration values computed. Default is :code:`1e-5` (cf.
        :func:`~gudhi.DelaunayComplex.get_float_relative_precision`). For more details, please refer to
        `CGAL::Lazy_exact_nt<NT>::set_relative_precision_of_to_double <https://doc.cgal.org/latest/Number_types/classCGAL_1_1Lazy__exact__nt.html>`_

:raises ValueError: If precision is not in (0, 1).
        )doc");
}

//
// _delaunay_complex.cc ends here
